apps/consulting/posts/improvements-to-bsr-matrix-multiplication-in-pytorch.md

```diff
@@ -0,0 +1,292 @@
+/>
+
+Comparison to the [dense baseline][#fig-5] is also very promising. Block sizes of 32 and 64 are
+faster than dense for all sparsity ratios, and block size 16 crossing the 1x boundary at 70% sparsity.
+This also represents a leap forward in the usability of sparse tensors. 
+
+<img src="posts/improvements-to-bsr-matrix-multiplication-in-pytorch/bsr_dense_mm_float32_vs_dense.png"
+    name="fig-5"
+    alt="A plot displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup over
+    the dense baseline on the vertical.  There is a horizontal line marking the speedup ratio of 1.0
+    above which the sparse subject takes less time to execute than the dense counterpart.  Three lines are
+    plotted indicating block sizes used for the sparse tensor of 16, 32, and 64."
+/>
+
+Now we look at the performance for half precision data types. The new kernel provides between 20-50x
+speedup over the [older sparse implementation][#fig-6]. While a speedup of this size is always a notable
+result, we have known that the implementation supporting half precision data types would not be
+optimal. However, as cuSPARSE did not support these data types a working implementation was created
+by composing calls into the dense kernel for each sparse block. This approach suffers from a
+significant penalty due to launching many CUDA kernels. 
+
+<img src="posts/improvements-to-bsr-matrix-multiplication-in-pytorch/bsr_dense_mm_half_vs_sparse.png"
+    name="fig-6"
+    alt="A plot displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup over
+    the older sparse implementation on the vertical.  Three lines are plotted indicating block sizes
+    used for the sparse tensor of 16, 32, and 64."
+/>
+
+Comparing performance for half precision to the [dense baseline][#fig-7] we again see one clear win.
+Using a block size of 64 the triton implementation begins to out perform the dense matrix multiply
+at 60% sparsity.  Unfortunately, this does not hold for smaller block sizes with 32 reach >1x
+speedup at 90% sparsity, and block size 16 only passing this point at 99%. 
+
+<img src="posts/improvements-to-bsr-matrix-multiplication-in-pytorch/bsr_dense_mm_half_vs_dense.png"
+    name="fig-7"
+    alt="A plot displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup over
+    the older sparse implementation on the vertical.  There is a horizontal line marking the speedup
+    ratio of 1.0 above which the sparse subject takes less time to execute than the dense
+    counterpart.  Three lines are plotted indicating block sizes used for the sparse tensor of 16,
+    32, and 64."
+/>
+
+## Try it out!
+
+These kernels will be shipped with PyTorch 2.1. They can all be found in the
+`torch.sparse._triton_ops` private submodule. Although not yet stable, the curious user may want
+experiment with these tools. If you want to stick to the public features only, the `bsr_dense_mm`
```

Comment:
The title of this section is "try it out". Then it goes into discussing where to find these, and then comes back to the "try it out" topic. It'd be better to start noting that this is not stable, and then go onto discussing where can you find it.